pkg/server/plugin/keymanager/awskms/awskms.go

```diff
@@ -0,0 +1,610 @@
+			AliasName:   aws.String(entry.Alias),
+			TargetKeyId: &entry.KMSKeyID,
+		})
+		if err != nil {
+			return newErrorf(codes.Internal, "failed to update alias: %v", err)
+		}
+		p.log.Debug("Alias updated", aliasTag, entry.Alias)
+
+		select {
+		case p.scheduleDelete <- oldEntry.KMSKeyID:
+			p.log.Debug("Key enqueued for deletion", keyIDTag, oldEntry.KMSKeyID)
+		default:
+			p.log.Debug("Failed to enqueue key for deletion", keyIDTag, oldEntry.KMSKeyID)
+		}
+	}
+	return nil
+}
+
+func (p *Plugin) setCache(keyEntries []*keyEntry) {
+	// clean previous cache
+	p.entries = make(map[string]keyEntry)
+
+	// add results to cache
+	for _, e := range keyEntries {
+		p.entries[e.PublicKey.Id] = *e
+		p.log.Debug("Key loaded", keyIDTag, e.KMSKeyID, aliasTag, e.Alias)
+	}
+}
+
+// scheduleDeleteTask ia a long running task that deletes keys that were rotated
+func (p *Plugin) scheduleDeleteTask(ctx context.Context) {
+	backoffMin := 1 * time.Second
+	backoffMax := 60 * time.Second
+	backoff := backoffMin
+
+	for {
+		select {
+		case <-ctx.Done():
+			p.log.Debug("Stopping schedule delete task", "reason", ctx.Err())
+			p.notifyDelete()
+			return
+		case keyID := <-p.scheduleDelete:
+			log := p.log.With(keyIDTag, keyID)
+			_, err := p.kmsClient.ScheduleKeyDeletion(ctx, &kms.ScheduleKeyDeletionInput{
+				KeyId:               aws.String(keyID),
+				PendingWindowInDays: aws.Int32(7),
+			})
+
+			if err == nil {
+				log.Debug("Key deleted")
+				backoff = backoffMin
+				p.notifyDelete()
+				continue
+			}
+
+			var notFoundErr *types.NotFoundException
+			if errors.As(err, &notFoundErr) {
+				log.Error("No such key, dropping from delete schedule")
+				p.notifyDelete()
+				continue
+			}
+
+			var invalidArnErr *types.InvalidArnException
+			if errors.As(err, &invalidArnErr) {
+				log.Error("Invalid ARN, dropping from delete schedule")
+				p.notifyDelete()
+				continue
+			}
+
+			log.Error("It was not possible to schedule key for deletion", "reason", err)
+			select {
+			case p.scheduleDelete <- keyID:
+				log.Debug("Key re-enqueued for deletion")
+			default:
+				log.Error("Failed to re-enqueue key for deletion")
+			}
+			p.notifyDelete()
+			backoff = min(backoff*2, backoffMax)
+			time.Sleep(backoff)
+		}
+	}
+}
+
+func (p *Plugin) aliasFromSpireKeyID(spireKeyID string) string {
+	return aliasPrefix + p.keyPrefix + spireKeyID
+}
+
+func (p *Plugin) descriptionFromSpireKeyID(spireKeyID string) string {
+	return p.keyPrefix + spireKeyID
+}
+
+func (p *Plugin) notifyDelete() {
+	if p.hooks.deleteSignal != nil {
+		p.hooks.deleteSignal <- struct{}{}
+	}
+}
+
+type keyFetcher struct {
+	log       hclog.Logger
+	kmsClient kmsClient
+	keyPrefix string
+}
+
+func (kf *keyFetcher) fetchKeyEntries(ctx context.Context) ([]*keyEntry, error) {
+	var keyEntries []*keyEntry
+	var keyEntriesMutex sync.Mutex
+	paginator := kms.NewListAliasesPaginator(kf.kmsClient, &kms.ListAliasesInput{})
+	g, ctx := errgroup.WithContext(ctx)
+
+	for {
```

Comment:
I haven't been able to find a better way to do this, however I thought I'd point out that this might generate a significant number of back-to-back API calls, particularly on large AWS accounts with heavy KMS usage. This could become a problem as these calls get ratelimited and/or rejected, causing SPIRE Server to fail boot. Anything we can do to reduce the call overhead (e.g. large page size, more targeted query, etc) will be worth it.

Reply:
I just changed the page size to 100 which is the max. The API doesn't offer much more than that.